extra_sdk_options.go

```diff
@@ -24,11 +39,135 @@ func WithTenant(input string) (SDKOption, error) {
-		return WithTenantDomain(parts[0]), nil
+		normalize.useWithTenant = true
+		normalize.Tenant = parts[0]
+
+		return normalize, nil
 	}
 
 	u.Scheme = "https"
-	return WithServerURL(u.String()), nil
+	normalize.useWithServer = true
+	normalize.ServerURL = u.String()
+	return normalize, nil
+}
+
+func WithTenant(input string) (SDKOption, error) {
+	resp, err := normalizeTenant(input)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.useWithTenant {
+		return WithTenantDomain(resp.Tenant), nil
+	}
+
+	if resp.useWithServer {
+		return WithServerURL(resp.ServerURL), nil
+	}
+
+	return func(api *ConductoroneAPI) {}, nil
+}
+
+type CustomSDKOption func(*CustomOptions)
+
+//func WithClient(client HTTPClient) SDKOption {
+//	return func(sdk *ConductoroneAPI) {
+//		sdk.sdkConfiguration.DefaultClient = client
+//	}
+//}
+
+func WithTenantCustom(input string) (CustomSDKOption, error) {
+	resp, err := normalizeTenant(input)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(sdk *CustomOptions) {
+		sdk.useWithServer = resp.useWithServer
+		sdk.useWithTenant = resp.useWithTenant
+		sdk.ServerURL = resp.ServerURL
+		sdk.Tenant = resp.Tenant
+	}, nil
+}
+
+func WithLog(logger *zap.Logger) CustomSDKOption {
+	return func(sdk *CustomOptions) {
+		sdk.logger = logger
+	}
+}
+
+func WithUserAgent(userAgent string) CustomSDKOption {
+	return func(sdk *CustomOptions) {
+		sdk.userAgent = userAgent
+	}
+}
+func WithTLSConfig(tlsConfig *tls.Config) CustomSDKOption {
+	return func(sdk *CustomOptions) {
+		sdk.tlsConfig = tlsConfig
+	}
+}
+
+type CustomOptions struct {
+	useWithServer bool
+	useWithTenant bool
+
+	ServerURL string
+	Tenant    string
+
+	withClient *http.Client
+	logger     *zap.Logger
+	tlsConfig  *tls.Config
+
+	userAgent string
+}
+
+func NewWithCredentials(ctx context.Context, cred *ClientCredentials, opts ...CustomSDKOption) (*ConductoroneAPI, error) {
+	tokenSource, err := NewC1TokenSource(ctx, cred.ClientID, cred.ClientSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	options := &CustomOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if options.userAgent == "" {
+		options.userAgent = "conductorone-sdk-go"
+	}
+
+	uclient, err := uhttp.NewClient(
+		ctx,
+		uhttp.WithTokenSource(tokenSource),
+		uhttp.WithLogger(options.logger != nil, options.logger),
+		uhttp.WithUserAgent(options.userAgent),
+		uhttp.WithTLSClientConfig(options.tlsConfig),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	sdkOpts := []SDKOption{}
+	if options.useWithServer {
+		sdkOpts = append(sdkOpts, WithServerURL(options.ServerURL))
+	}
+	if options.useWithTenant {
+		sdkOpts = append(sdkOpts, WithTenantDomain(options.Tenant))
+	}
+	sdkOpts = append(sdkOpts, WithClient(uclient))
+
+	return New(sdkOpts...), nil
+}
+
+func NewWithLogin(ctx context.Context, tenantName string, clientID string, personalClientCredentialDisplayName string, cb func(validateUrl string) error, opts ...CustomSDKOption) (*ConductoroneAPI, error) {
```

Comment:
i'm not sure this makes sense?

Login() is a setup flow thing, not an every run thing?

Reply:
Works for me